### ISSUE TITLE
$JAVA_HOME is set with java-8-oracle on the SUDO-ENABLED environment

### ISSUE THREAD

**Issue by yatch** [travis-ci/travis-ci#6983]
# Summary
`$JAVA_HOME` points to the directory of `java-8-oracle` on the SUDO-ENABLED environment while `java --version` says it's 1.7.0_80:

```shell
$ java -version
java version "1.7.0_80"
Java(TM) SE Runtime Environment (build 1.7.0_80-b15)
Java HotSpot(TM) 64-Bit Server VM (build 24.80-b11, mixed mode)

$ echo $JAVA_HOME
/usr/lib/jvm/java-8-oracle
```

# Workaround
Add `group: deprecated` in your .travis.yml.
```YAML
group: deprecated
```

# More Info
An example job URL is https://travis-ci.org/yatch/travis-jdk7-test/builds/181735078.

The following is the contents of .travis.yml to reproduce this issue.
```YAML
language: java
sudo: required
script:
  - java -version
  - echo $JAVA_HOME
```

I ran `java --version` and `echo $JAVA_HOME` on other environments (no `dist` specified). Here is the summary:

|group|environment|java|$JAVA_HOME|note|
|-|-|-|-|-|
|stable|CONTAINER-BASED|"1.7.0_76"|/usr/lib/jvm/java-7-oracle|no group tag, no sudo tag|
|stable|SUDO-ENABLED|"1.7.0_80"|/usr/lib/jvm/java-8-oracle|sudo: required|
|deprecated|SUDO-ENABLED|"1.7.0_80"|/usr/lib/jvm/java-7-oracle|group: deprecated, sudo: enabled|

~~I suspect this is related to https://github.com/travis-ci/travis-ci/issues/6928.., but I'm not sure...~~ (edited)

**Comment by meatballhat** [comment]
@yatch Thanks for the issue!  Do you have any job URLs you can share?  Is the above `.travis.yml` the entire content?

**Comment by yatch** [comment]
@meatballhat Yes, please see this one: https://travis-ci.org/yatch/travis-jdk7-test/builds/181735078 . That is the entire content of `.travis.yml` used in the job for https://github.com/yatch/travis-jdk7-test/commit/d3b17c760ab7acf1ed8cb291fd1f4d74fa411135.

**Comment by meatballhat** [comment]
@yatch Thank you, that's super helpful!  I can confirm that this issue exists on `dist: precise`, but *does not* on `dist: trusty`, which means that it is a separate issue than what was addressed with the [build env update pushed today](https://docs.travis-ci.com/user/build-environment-updates/2016-12-06/).

- `dist: precise`: https://travis-ci.org/meatballhat/cloaked-octo-tyrion/builds/181759598
- `dist: trusty`: https://travis-ci.org/meatballhat/cloaked-octo-tyrion/builds/181760381

**Comment by yatch** [comment]
@meatballhat Wow, what a fast response...! Thank you for the confirmation. :-)

**Comment by meatballhat** [comment]
@yatch I should have asked above: Are you currently blocked by this?

**Comment by yatch** [comment]
No; I applied `group: deprecated` as a workaround, that I've learned this week ;-)

**Comment by offa** [comment]
I have the same issue on **_trusty_** (**container based**) with following `.travis.yml`:

```yml
language: java

dist: trusty

jdk:
    - openjdk7
    - oraclejdk7
    - oraclejdk8

cache:
    directories:
         - $HOME/.m2
```

Both JDK7 builds fail, but JDK8 is fine. Here's an excerpt from a [failed build](https://travis-ci.org/offa/NBCndUnit/jobs/182341378):

```
$ jdk_switcher use openjdk7

Switching to OpenJDK7 (java-1.7.0-openjdk), JAVA_HOME will be set to /usr/lib/jvm/java-7-openjdk

sudo: /usr/bin/sudo must be owned by uid 0 and have the setuid bit set 

[…]
$ java -Xmx32m -version

java version "1.8.0_111"

[…]
$ mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V

Error: JAVA_HOME is not defined correctly.

  We cannot execute /usr/lib/jvm/java-7-openjdk/bin/java

The command "eval mvn install -DskipTests=true -Dmaven.javadoc.skip=true -B -V" failed. Retrying, 2 of 3.
```

Restarting the failed builds and cleaning the cache didn't work so far.
Curiously another project with exactly the same `.travis.yml` passed.

**Comment by yatch** [comment]
@meatballhat I'm happy to know your patch, https://github.com/travis-ci/packer-templates/pull/371 , has been merged! Thank you!!

The next "build-environment-update" will resolve this issue, won't it? This issue ticket has been closed, but the issue itself seems to still remains on the system, https://travis-ci.org/yatch/travis-jdk7-test/builds/181748984 .

**Comment by meatballhat** [comment]
@yatch that's correct, the issue will be present until the next build env update, so I'm reopening.  Sorry about the auto-close.

**Comment by yatch** [comment]
@meatballhat Thanks for re-opening this issue. And, thanks to you, I've learned the "auto-close" feature via Pull-Request!! Have a nice weekend :-)

**Comment by offa** [comment]
Is there are workaround for sudo disabled trusty builds? Adding `group: deprecated` still fails the JDK7 builds.

**Comment by meatballhat** [comment]
@offa the container-based images are behind the sudo-enabled ones, sorry about that.  Are you able to set an explicit jdk key like so?:

``` yaml
jdk: openjdk7
```

**Comment by offa** [comment]
> @offa the container-based images are behind the sudo-enabled ones, sorry about that.

I see, that's probably why the problem doesn't exist if `sudo: required` is used (#7019)

> Are you able to set an explicit jdk key like so?:

I'll give it a try, one moment.
